### ISSUE TITLE
support local files in npm-shrinkwrap

### ISSUE THREAD

**Issue by tomyan** [npm/npm#3377]
I'd like to be able to replace the resolved urls in npm-shrinkwrap with local files (possibly file:// uris), so that I can rebuild my dependencies without needing to use the public npm repo. I've read this blogpost, which explains the approach of just checking in the names and versions of the modules rather than the binaries:

http://blog.nodejs.org/2012/02/27/managing-node-js-dependencies-with-shrinkwrap/

However, I'm trying to build a source rpm that has no external dependencies (other than those declared as build dependencies), that can be used to build architecture/abi specific rpms (specifically without needing to call out over the network). Happy to work on a pull request if this seems like a valuable feature.

Thanks

Tom

**Comment by zkat** [comment]
Hey @tomyan

The idea of a local cache is really interesting to us, and goes well with our current plans to work on how shrinkwrap and caching in general work. @othiym23 moved that particular feature request over to #12443.

As far as addressing your specific request here, we believe that everything you need in order to support cross-platform native builds is already available to you:

Tl;dr:
1. Run your initial npm install in your local dev machine, with `--ignore-scripts`.
2. Have your rpm build run `npm rebuild` after grabbing the package on your target platform. This will both build native modules, *and* run any relevant lifecycle scripts on the package and its dependencies.
3. Build the rpm, including `node_modules/`, as you usually would.

The above should execute without any network access, unless your dependencies’ lifecycle scripts happen to do networking stuff (though cache changes would not help you at all here).

Since I think the above addresses your stated needs, I’m going to close this issue. Please feel free to comment on this issue if you feel the above doesn’t actually help, and thanks for your time in submitting this. I know it’s been a while by now. ;)

**Comment by JamieMason** [comment]
Hi all,
I just stumbled on this thread and thought I'd mentioned that I've been working on something along these lines in the form of https://github.com/JamieMason/shrinkpack, hopefully something in there might be useful to you.

Thanks.
